k8s/apis/k8s/v1alpha1/sidecar_types.go

```diff
@@ -0,0 +1,177 @@
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:="IfNotPresent"
+	ImagePullPolicy string `json:"imagePullPolicy,omitempty"`
+}
+
+// VolumeConfig holds all configuration about persistent volume
+type VolumeConfig struct {
+	// the name of pvc
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=""
+	PvcName string `json:"pvcName,omitempty"`
+
+	// the mount path of pv
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=""
+	MountPath string `json:"mountPath,omitempty"`
+}
+
+// SidecarSpec defines the desired state of Sidecar
+type SidecarSpec struct {
+	// the image of vineyard sidecar image
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:="vineyardcloudnative/vineyardd:latest"
+	Image string `json:"image,omitempty"`
+
+	// the policy about pulling vineyard sidecar image
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:="IfNotPresent"
+	ImagePullPolicy string `json:"imagePullPolicy,omitempty"`
+
+	// the selector of pod
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=""
+	Selector string `json:"selector,omitempty"`
+
+	// the replicas of workload
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=0
+	Replicas int `json:"replicas,omitempty"`
+
+	// shared memory size for vineyardd
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:="256Mi"
+	Size string `json:"size,omitempty"`
+
+	// The directory on host for the IPC socket file. The UNIX-domain
+	// socket will be placed as `${Socket}/vineyard.sock`.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:="/var/run"
+	Socket string `json:"socket,omitempty"`
+
+	// memory threshold of streams (percentage of total memory)
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:=80
+	StreamThreshold int64 `json:"streamThreshold,omitempty"`
+
+	// path of etcd executable
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:="etcd"
+	EtcdCmd string `json:"etcdCmd,omitempty"`
+
+	// endpoint of etcd
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:="http://etcd-for-vineyard:2379"
+	EtcdEndpoint string `json:"etcdEndpoint,omitempty"`
+
+	// path prefix in etcd
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:="/vineyard"
+	EtcdPrefix string `json:"etcdPrefix,omitempty"`
+
+	// the configuration of spilling
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={}
+	Spill Spill `json:"spill,omitempty"`
+
+	// metric configuration
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={}
+	Metric Metric `json:"metric,omitempty"`
+
+	// metric configuration
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={}
+	Volume VolumeConfig `json:"volume,omitempty"`
+
+	// rpc service configuration
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={type: "ClusterIP", port: 9600, selector: "rpc.vineyardd.v6d.io/rpc=vineyard-rpc"}
+	Service ServiceConfig `json:"service,omitempty"`
+}
```

Comment:
Looks duplicate a lot with `vineyardd_types.go`...

Reply:
No problem, I will add the doc in the next pr. Yeah, the Sidecar is much like Vineyardd, it is used to provide observability, showing users the detailed configuration of the vineyard after injection.

Reply:
Could you please try to make those duplicated structs into somewhere like `common.go` or `utils.go`? (I'm not sure whether kubebuilder can handle that.

Reply:
> I will add the doc in the next pr.

Why not in this pull request as well? I was imaging the feature and corresponding documentation should be treat as a whole.

Reply:
OK, If this doesn't burden your review, I'm willing to add it in this pr.

Reply:
> Could you please try to make those duplicated structs into somewhere like `common.go` or `utils.go`? (I'm not sure whether kubebuilder can handle that.

Thanks for the advice, a possible way is to integrate some configurations of vineyard(image, command, etc) into a structure, store it in `vineyardd_types.go`, and then reuse this structure in `sidecar_types.go`, what do you think?